subgraphs/_reference_/src/sdk/protocols/bridge/pool.ts

```diff
@@ -0,0 +1,638 @@
+import { TokenManager } from "./tokens";
+import { PoolSnapshot } from "./pool_snapshot";
+import { SDK } from ".";
+
+type onCreatePoolCallback<T> = (
+  event: ethereum.Event,
+  pool: Pool,
+  sdk: SDK,
+  aux: T | null
+) => void;
+
+export class PoolManager {
+  protocol: Bridge;
+  tokens: TokenManager;
+
+  constructor(protocol: Bridge, tokens: TokenManager) {
+    this.protocol = protocol;
+    this.tokens = tokens;
+  }
+
+  loadPool<T>(
+    id: Bytes,
+    onCreate: onCreatePoolCallback<T> | null = null,
+    aux: T | null = null
+  ): Pool {
+    let entity = PoolSchema.load(id);
+    if (entity) {
+      return new Pool(this.protocol, entity, this.tokens);
+    }
+
+    entity = new PoolSchema(id);
+    entity.protocol = this.protocol.getBytesID();
+
+    const pool = new Pool(this.protocol, entity, this.tokens);
+    pool.isInitialized = false;
+    if (onCreate) {
+      onCreate(this.protocol.getCurrentEvent(), pool, this.protocol.sdk!, aux);
+    }
+    return pool;
+  }
+}
+
+export class Pool {
+  pool: PoolSchema;
+  protocol: Bridge;
+  tokens: TokenManager;
+  snapshoter: PoolSnapshot | null = null;
+
+  public isInitialized: boolean = true;
+
+  constructor(protocol: Bridge, pool: PoolSchema, tokens: TokenManager) {
+    this.pool = pool;
+    this.protocol = protocol;
+    this.tokens = tokens;
+    this.snapshoter = new PoolSnapshot(pool, protocol.event);
+  }
+
+  private save(): void {
+    this.pool.save();
+  }
+
+  private getInputToken(): Token {
+    return this.tokens.initToken(Address.fromBytes(this.pool.inputToken));
+  }
+
+  initialize(
+    name: string,
+    symbol: string,
+    type: BridgePoolType,
+    inputToken: Token
+  ): void {
+    if (this.isInitialized) {
+      return;
+    }
+
+    const event = this.protocol.getCurrentEvent();
+    this.pool.name = name;
+    this.pool.symbol = symbol;
+    this.pool.type = type;
+    this.pool.inputToken = inputToken.id;
+    this.pool.destinationTokens = [];
+    this.pool.routes = [];
+    this.pool.createdTimestamp = event.block.timestamp;
+    this.pool.createdBlockNumber = event.block.number;
+
+    if (type == BridgePoolType.BURN_MINT) {
+      this.pool.mintSupply = BIGINT_ZERO;
+    }
+    this.pool.inputTokenBalance = BIGINT_ZERO;
+    this.pool.totalValueLockedUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeSupplySideRevenueUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeProtocolSideRevenueUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeTotalRevenueUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeVolumeIn = BIGINT_ZERO;
+    this.pool.cumulativeVolumeOut = BIGINT_ZERO;
+    this.pool.netVolume = BIGINT_ZERO;
+    this.pool.cumulativeVolumeInUSD = BIGDECIMAL_ZERO;
+    this.pool.cumulativeVolumeOutUSD = BIGDECIMAL_ZERO;
+    this.pool.netVolumeUSD = BIGDECIMAL_ZERO;
+
+    this.pool._lastDailySnapshotTimestamp = BIGINT_ZERO;
+    this.pool._lastHourlySnapshotTimestamp = BIGINT_ZERO;
+    this.save();
```

Comment:
Yeah 🤔 
Now it just calls `entity.save()`, but before it would also take care of updating the snapshots.

It’s also a place where to add any hooks that we want to run on save. Or if these wrappers managed more than 1 entity at a time you’d only need one save. It’s hypotheticals though